### PR TITLE
Issue 2490: Enable textual/base64 representation of a StreamCut.

### LINKED ISSUE #2490: Enable textual representation of a StreamCut.
**Problem description**
Enable a textual representation of a StreamCut which can also be used to reconstruct the object.
This textual representation can be used as a command line parameter while starting the connector job.

E.g:
` flink run MyJob --input scope1/stream1 --from <cut-data>`

**Problem location**
StreamCut.

**Suggestions for an improvement**
At present `StreamCut` supports `java.io.Serializable` java serialization;  Ensure we have a textual representation of it

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
Enable a textual/base64 representation of a StreamCut which can also be used to reconstruct the Streamcut object

**Purpose of the change**  
Fixes #2490 

**What the code does**  
Enables a textual representation of a StreamCut. This will enable a simpler way to pass StreamCut to connectors as a program parameter.

    `StreamCut#asText()` will return the textual/base64 representation of a StreamCut.
    `StreamCut#from(String)` will construct the StreamCut object from its base64 representation.

E.g:
StreamCut#toString() `scope/stream:4294967300=40, 4294967298=20, 4294967299=30, 4294967297=10`
StreamCut#asText() `H4sIAAAAAAAAACtOzi9I1S8uKUpNzLUyMbI0sTQzNzYwsDUx0FGAco0sLWyNkLmWtsbIXHNbQwMA/wWKVUcAAAA=`

**How to verify it**  
All existing tests should pass, new tests have been added to cover the new API

### CODE REVIEW

File: client/src/main/java/io/pravega/client/stream/StreamCut.java [hunk 1]
```diff
@@ -67,4 +84,15 @@ static StreamCut fromBytes(ByteBuffer cut) {
...
         return StreamCutInternal.fromBytes(cut);
     }
+
+    /**
+     * Deserializes a Base64 representation of the StreamCut obtained via {@link StreamCut#asText()}.
+     * @param base64String Base64 representation of StreamCut obtained using {@link StreamCut#asText()}
+     * @return The StreamCut object
+     * @throws SerializationException Exception during deserialization.
+     */
+    static StreamCut of(String base64String) throws SerializationException {
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/stream/StreamCut.java]: please consider renaming this method to `from`

File: client/src/main/java/io/pravega/client/stream/StreamCut.java [hunk 2]
```diff
@@ -54,6 +64,13 @@ private Object readResolve() {
      * Serializes the cut to a compact byte array.
      */
     ByteBuffer toBytes();
+
+    /**
+     * Serializes the cut to a compact base64 string representation.
+     * @return Base64 representation of the StreamCut.
+     * @throws SerializationException Exception during serialization.
+     */
+    String asText() throws SerializationException;
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/stream/StreamCut.java]: the implementation of `asText` is left to implementer of this interface while `of` is implemented here. 
Both logic should coexist.

Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/StreamCut.java]: done.

File: client/src/main/java/io/pravega/client/stream/StreamCut.java (hunk 1 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/StreamCut.java]: corrected it.

File: client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java [hunk 3]
```diff
@@ -127,7 +130,12 @@ public ByteBuffer toBytes() {
         ByteArraySegment serialized = SERIALIZER.serialize(this);
         return ByteBuffer.wrap(serialized.array(), serialized.arrayOffset(), serialized.getLength());
     }
-    
+
+    @Override
+    public String asText() throws SerializationException {
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java]: please remove it from here as its now implemented in the StreamCut interface

Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java]: done.

**Comment by EronWright** [comment]
@shrids please update the description with some examples, thanks

File: common/src/main/java/io/pravega/common/util/ToStringUtils.java [hunk 4]
```diff
@@ -63,4 +74,43 @@
...
+     * Get the original string from its compressed base64 representation.
+     * @param base64CompressedString Compressed Base64 representation of the string.
+     * @return The original string.
+     * @throws IOException If an I/O exception occurs.
+     * @throws NullPointerException If base64CompressedString is null.
+     * @throws IllegalArgumentException If base64CompressedString is not null, but has a length of zero or if the string has illegal base64 character.
+     */
+    public static String decompressFromBase64(final String base64CompressedString) throws IOException {
+        Exceptions.checkNotNullOrEmpty(base64CompressedString, "base64CompressedString");
+        byte[] dataBytes = Base64.getDecoder().decode(base64CompressedString);
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/ToStringUtils.java]: From what I see, `getDecoder` and `getEncoder` can wrap existing `InputStreams`, which (I hope) will save you from having to allocate the extra array, and thus passing through it twice.

Please check it out and, if so, change it both here and in `compressToBase64` above.

Comment (shrids) [reply common/src/main/java/io/pravega/common/util/ToStringUtils.java]: Yes, wrapping a stream with `getDecoder().wrap(..)` and `getEncoder().wrap(..)` is elegant and avoids using an extra array.
Updated the `compressToBase64` and `decompressFromBase64` to use it.

**Comment by EronWright** [comment]
Does the string contain a version number, so that we may evolve the algorithm?

I think the textual representation chould be made significantly more compact with a better algorithm.   I would challenge the value of GZIP here, and suggest taking advantage of the fact that we're dealing with numbers with known semantics.  

For example, consider that a cut may be represented by a list of tuples `(epoch, segmentNumber, offset)`.
Unzip the tuples into three numeric arrays of equal length, convert each array to a byte array, then base64-encode.   RLE or delta encoding may also be useful, especially for epoch (I assume).  Note that you may sort the tuples for additional benefit.
```
int[] epochs; int[] segmentNumbers; long[] offsets;
return base64(version + length + encodeEpochs(epoch)) + encodeSegments(segmentNumbers) + encodeOffsets(offsets));
```

**Comment by shrids** [comment]
Thanks @EronWright for taking a look at the PR.
The intent of adding gzip was to compress the data and also to ensure partial data is detected.
Will try out the mechanism suggested by you.
Comparison of the currently evaluated text representations: 
```
ToString                           : scope/stream:4294967300=20, 8589934600=50, 4294967298=20, 8589934601=60, 4294967299=30, 4294967297=10, 8589934597=50
ToStringBase64                     : c2NvcGUvc3RyZWFtOjQyOTQ5NjczMDA9MjAsIDg1ODk5MzQ2MDA9NTAsIDQyOTQ5NjcyOTg9MjAsIDg1ODk5MzQ2MDE9NjAsIDQyOTQ5NjcyOTk9MzAsIDQyOTQ5NjcyOTc9MTAsIDg1ODk5MzQ1OTc9NTA=
ToString_withGzipBase64            : H4sIAAAAAAAAACtOzi9I1S8uKUpNzLUyMbI0sTQzNzYwsDUy0FGwMLWwtDQ2MQNyTYFcqKyRpQWqrKGtGbKspa0xMtfc1hCh2BTINTUAAKFIfE90AAAA
VersionedSerializer_Base64         : rO0ABXNyADppby5wcmF2ZWdhLmNsaWVudC5zdHJlYW0uaW1wbC5TdHJlYW1DdXRJbXBsJFNlcmlhbGl6ZWRGb3JtAAAAAAAAAAECAAFbAAV2YWx1ZXQAAltCeHB1cgACW0Ks8xf4BghU4AIAAHhwAAAARAAAAQAAAAA8AAxzY29wZS9zdHJlYW0FwAAAAQAAAAEKwAAAAQAAAAIUwAAAAQAAAAMewAAAAQAAAAQUwAAAAgAAAAUy
VersionedSerializer_withGzipBase64 : H4sIAAAAAAAAAFvzloG1uIjBKjNfr6AosSw1PVEvOSczNa9Er7ikKDUxVy8ztyBHLxjMdi4t8QTyVIJTizITczKrUlPc8otyGSCAkYmBMZqBtSwxpzS1hIEp2qmioLQIRK/5LP6DjSPkARMDQ0UBUKELUC1Igw0DT3FyfkGqPsQi1gMQcUYuKINJBMpgloMyWEAiQGMYWI0ALqRsjLcAAAA=

The PR current implements "ToString_withGzipBase64 "
```

File: client/src/main/java/io/pravega/client/stream/StreamCut.java [hunk 5]
```diff
@@ -54,10 +65,34 @@ private Object readResolve() {
...
      */
     ByteBuffer toBytes();
-    
+
+    /**
+     * Obtains the compact base64 string representation of StreamCut.
+     *
+     * @return Base64 representation of the StreamCut.
+     */
+    @SneakyThrows(IOException.class)
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/StreamCut.java]: done.

File: client/src/main/java/io/pravega/client/stream/StreamCut.java [hunk 6]
```diff
@@ -54,10 +65,34 @@ private Object readResolve() {
...
+        return compressToBase64(this.toString());
+    }
+
+    /**
+     * Obtains the a StreamCut object from its Base64 representation obtained via {@link StreamCut#asText()}.
+     *
+     * @param base64String Base64 representation of StreamCut obtained using {@link StreamCut#asText()}
+     * @return The StreamCut object
+     */
+    @SneakyThrows(IOException.class)
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/StreamCut.java]: done.

File: common/src/main/java/io/pravega/common/util/ToStringUtils.java [hunk 7]
```diff
@@ -63,4 +76,43 @@
...
     }
 
+    /**
+     * Convert the given string to its compressed base64 representation.
+     * @param string String to be compressed to base64.
+     * @return String Compressed Base64 representation of the input string.
+     * @throws IOException If an I/O exception occurs.
+     * @throws NullPointerException If string is null.
+     */
+    public static String compressToBase64(final String string) throws IOException {
```
[the remark this answers is not in the document]
Comment (shrids) [reply common/src/main/java/io/pravega/common/util/ToStringUtils.java]: removed it.

File: common/src/main/java/io/pravega/common/util/ToStringUtils.java [hunk 8]
```diff
@@ -63,4 +76,43 @@
...
+
+    /**
+     * Get the original string from its compressed base64 representation.
+     * @param base64CompressedString Compressed Base64 representation of the string.
+     * @return The original string.
+     * @throws IOException If an I/O exception occurs.
+     * @throws NullPointerException If base64CompressedString is null.
+     * @throws IllegalArgumentException If base64CompressedString is not null, but has a length of zero.
+     */
+    public static String decompressFromBase64(final String base64CompressedString) throws IOException {
```
[the remark this answers is not in the document]
Comment (shrids) [reply common/src/main/java/io/pravega/common/util/ToStringUtils.java]: yes, it should thrown an `IllegalArgumentException`  and not an `IOException` if 
- input is not a base64 encoded string.
- input has partial data.

Updated the test code to validate this behavior.

File: client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java [hunk 9]
```diff
@@ -64,16 +69,28 @@ public Stream getStream() {
     public StreamCutInternal asImpl() {
         return this;
     }
-    
+
     @Override
     public String toString() {
-        return stream.getScopedName() + ":"
+        return stream.getScopedName() + ":" + TO_STRING_VERSION + ":"
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java]: Let's not overload toString. This really should be for debugging. If we need to create a method that is going to be versioned for serialization, I think it should called something else.

Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java]: done.

**Comment by shrids** [comment]
I have added a version to the textual representation, which can be used to improve the algorithm in future.	
	
Compared the approaches:

Approach 1: GZIP the StreamCut for a compact textual representation
Approach 2: Use Delta encoding for segment numbers and offsets and RLE for epochs with sorting.

The advantages of Approach 1 are:
```
1. Ability to handle corrupted data due to checksum.
2. Ability to handle partial data. ()
3. easy to debug `echo $streamCut | base64 -d | gunzip on bash prompt.
4. Easier to adapt when the StreamCut will have future segments / predecessor segments which will be added  as part of StreamCut validation task.
```
The advantage of Approach 2 is:
```
1. For the given data distribution (chosen in the example) the compression is around 36 % smaller. (121 to 77 character).
```

**Comment by EronWright** [comment]
Thanks for the investigation @shrids.  If we are being honest, some of the "advantages" of (1) seem cooked up to make a point. 
1. A checksum could be added to either solution.  
2. The phrase 'partial data' is undefined but the various meanings that I could divine seem solvable or invalid.   

Point (4) is troubling because it suggests that the streamcut is going to become more unwieldy to the task of being used as a parameter.   The predecessors and successors seem to be info that can be unambiguously looked up.

Meanwhile you've demonstrated that a more compact representation is possible.   Since you added a version field, I suppose we can move to a more compact representation later, if you don't have time to implement it now.

**Comment by shrids** [comment]
@EronWright Yes, Point (4) implies that either of the approaches would end up making the textual representation of StreamCut longer. 
Also,  with the compact representation (approach 2) the number of characters in the above example with `segment count less than 10 `is `77` , which is high for command line argument.  
Given that the number the segments in a real world scenario can be still higher both the approaches seem unwieldy.

**Comment by EronWright** [comment]
Since you kindly wrote some code already, how about we use a packed structure like I suggested.

**Comment by shrids** [comment]
Thanks for your feedback @EronWright . This PR can wait until Point 4 (additional predecessor/successor data will be part of the StreamCut)  is finalized. Will implement the changes with packed structure after point 4 is finalized in the interest of time.

**Comment by shrids** [comment]
Updated the code to reduce the length of the textual representation @EronWright 

The GZIPOutputStream is a wrapper around DEFLATE compression (`java.util.zip.DeflaterOutputStream`), which uses a form of `RunLengthEncoding` to remove redundancy.
The compression level with the latest code is similar to the packed structure with checksum.

File: client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java [hunk 10]
```diff
@@ -74,6 +87,50 @@ public String toString() {
                                                                                e -> e.getValue())));
     }
 
+    @Override
+    public String asText() {
+        return compressToBase64(getText());
```
Comment (fpj) [inline client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java]: Say we change the method we use to generate a textual representation in the future, how do we make the reconstruction compatible? Or do we simply throw an error at the application?

File: client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java [hunk 11]
```diff
@@ -74,6 +87,50 @@ public String toString() {
...
+
+        // append segmentsNumbers, epochs and offsets.
+        builder.append(listToString(segmentNumbers)).append(":");
+        builder.append(listToString(epochs)).append(":");
+        builder.append(listToString(offsets));
+
+        return builder.toString();
+    }
+
+    public static StreamCutInternal from(String textualRepresentation) {
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java]: Done, added the javadoc.

File: client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java (hunk 10 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java]: `StreamCutImpl#TO_STRING_VERSION` can be used to handle multiple versions of textual representation.

**Pull request merged**